### PR TITLE
Change ScrollIntoView to scroll more naturally

### PR DESCRIPTION
**Pull request by MrTheMake**
Video: https://streamable.com/1o1ms (Implementation in the playlist list follows later)

### CODE REVIEW

**Comment by peppy** [comment]
i like this one, nice

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 1]
```diff
@@ -311,7 +311,28 @@ private void scrollTo(float value, bool animated, double distanceDecay = float.P
...
+        /// <param name="animated">Whether to animate the movement.</param>
+        public void ScrollIntoView(Drawable d, bool animated = true)
+        {
+            float childPos = GetChildPosInContent(d);
+
+            // Scroll up or down depending on the current scroll position
+            if (childPos < Current)
+                ScrollTo(childPos, animated);
+            else if (childPos > Current + displayableContent - d.Height)
+                ScrollTo(childPos - displayableContent + d.Height, animated);
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: i'm pretty sure you want `d.DrawHeight`?

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 2]
```diff
@@ -64,6 +70,9 @@ public bool ScrollbarOverlapsContent
 
         private float displayableContent => ChildSize[scrollDim];
 
+        /// <summary>
+        /// Controls the scroll distance when scrolling with the mouse wheel.
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: Why not make the comments actually add something to the variable (your comment only makes things more ambiguous). This should mention that the distance added is per scroll "tick".

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 3]
```diff
@@ -279,6 +293,10 @@ protected override bool OnWheel(InputState state)
 
         private void onScrollbarMovement(float value) => scrollTo(clamp(value / scrollbar.Size[scrollDim]), false);
 
+        /// <summary>
+        /// Offsets the scroll position.
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: this is pointless

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 4]
```diff
@@ -114,6 +123,11 @@ public bool ScrollbarOverlapsContent
 
         protected override Container<Drawable> Content => content;
 
+        /// <summary>
+        /// Determines whether the available content is scrolled to the end.
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: what is "available content" it's not even publicly exposed.

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 5]
```diff
@@ -297,8 +315,18 @@ public void ScrollToEnd(bool animated = true, bool allowDuringDrag = false)
             if (!isDragging || allowDuringDrag) scrollTo(scrollableExtent, animated, DistanceDecayJump);
         }
 
+        /// <summary>
+        /// Scrolls by a relativ offset.
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: typo

also it's not "by a relative". relative has a different meaning for positioning.

File: osu.Framework/Graphics/Containers/ScrollContainer.cs (hunk 4 above)
Comment (MrTheMake) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: I thought it is the common term as it was used by `ScrollToEnd` already

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 6]
```diff
@@ -46,6 +49,9 @@ public bool ScrollbarVisible
 
         private bool scrollbarOverlapsContent = true;
 
+        /// <summary>
+        /// Determines whether the scrollbar overlaps the available content.
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: again, this should not be "available"

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 7]
```diff
@@ -114,6 +123,11 @@ public bool ScrollbarOverlapsContent
 
         protected override Container<Drawable> Content => content;
 
+        /// <summary>
+        /// Determines whether the content is scrolled to the end.
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: You should use "determines" less. You're using it both in places values can be set and places they can't be set. Personally I prefer it to only be used in get-only cases.

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 8]
```diff
@@ -297,8 +315,18 @@ public void ScrollToEnd(bool animated = true, bool allowDuringDrag = false)
             if (!isDragging || allowDuringDrag) scrollTo(scrollableExtent, animated, DistanceDecayJump);
         }
 
+        /// <summary>
+        /// Offsets the position.
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: Scroll to a new position relative to the current scroll offset.

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 9]
```diff
@@ -297,8 +315,18 @@ public void ScrollToEnd(bool animated = true, bool allowDuringDrag = false)
             if (!isDragging || allowDuringDrag) scrollTo(scrollableExtent, animated, DistanceDecayJump);
         }
 
+        /// <summary>
+        /// Offsets the position.
+        /// </summary>
+        /// <param name="offset">The scroll offset.</param>
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: The amount by which we should scroll.

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 10]
```diff
@@ -311,8 +339,34 @@ private void scrollTo(float value, bool animated, double distanceDecay = float.P
...
+
+            // Scroll up or down depending on the current scroll position
+            if (childPos < Current)
+                ScrollTo(childPos, animated);
+            else if (childPos > Current + displayableContent - d.DrawHeight)
+                ScrollTo(childPos - displayableContent + d.DrawHeight, animated);
+        }
 
+        /// <summary>
+        /// Determines the position (in relation to the current <see cref="scrollDim"/>) of a child in the content.
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: scrollDim is private

**Pull request merged**